vllm/worker/model_runner.py

```diff
@@ -18,7 +20,8 @@
 _PAD_SLOT_ID = -1
 # Capture graphs for batch size 1, 2, 4, 8, 16, 24, 32, 40, ..., 256.
 # NOTE: _get_graph_batch_size needs to be updated if this list is changed.
-_BATCH_SIZES_TO_CAPTURE = [1, 2, 4] + [8 * i for i in range(1, 33)]
+_BATCH_SIZES_TO_CAPTURE = [1, 2, 4, 8] #+ [8 * i for i in range(1, 33)]
+_BLOCK_COUNTS_TO_CAPTURE = [1, 2, 4] + [8 * i for i in range(1, 33)]
```

Comment:
I'm trying to understand what's the benefit of padding in seq_len dimension if we already have PA, which is naturally forcing sequence_lengths in attention to be rounded to block_size.

Let's say we have a block_size of N and we pad block counts to K. This is almost the same as if we had a block_size of N*K. What's more is that block_size=N*K is actually beneficial for us as it would reduce compilation times and improve block gathering performance as seen in PA traces. The only benefit I see for adding padding is that you could have finer control over the sequence_lengths like in the list above, where first block_counts are less then 8.

Reply:
> Let's say we have a block_size of N and we pad block counts to K. This is almost the same as if we had a block_size of NxK.

It's really not. It's just about limiting overcompute. Block size is still N. The only thing that actually changes with padding seq_len (or rather context_len) is how you pad the block table's second dimension. Cache itself, (and how we access it) is unaffected. Block table is going to be of dynamic shape (num_seqs, max_num_blocks_per_seq), with `max_num_blocks_per_seq` increasing  every `block_size` decode iterations. In eager mode (or, in HPU terms, in lazy PT backend), you'll see graph recompilation when that happens. In HPUGraphs/CUDAGraphs, you will not have any dynamicity, as  `max_num_blocks_per_seq` will never change and be set to total maximum context length // block size (e.g. for block size 16 and max_context_len 4096, you'll have 256 always). With CUDA kernels you can easily handle out of bound cases, but with Gaudi we always overcompute up to  `max_num_blocks_per_seq` determined by block table's second dimension. That means, if you're using HPUGraphs, regardless whether you have context length of let's say 20, it'll always cost as computing for max_context_len, e.g. 4096 - you will waste compute on computing 4076 entries in block table that you already have. Bucketing allows us to cap the upper limit at something lower. If you have context len 20, you'll hit a bucket with max_context_len of 32 tokens (16*2), doing overcompute of only 12 elements. Then, once you emit 33rd token, you'll transfer to a second bucket that can handle up to 64 tokens.

Reply:
https://github.com/HabanaAI/vllm-fork/blob/habana_main/vllm/hpu/ops.py#L33 - this part is the exact one that we want to limit; now with HPUGraphs we'll always have seq_len fixed at max value, and you'll have plenty of garbage data fetched and passed through GEMMs, contributing absolutely nothing to the end result

Reply:
Ok, and why can't the block_table grow by one, exactly how it grows in non-graph runs? What's more, with bucketing it also grows. The only difference is that it grows by chunks > 1 . 

I agree that computing up to whole max_context_len makes no sense, and I don't see a reason why we should always pad to it.

Let's work with your example. Assuming block_size=16:
seq_length=20 -> block_table=[bs, 2*16] utilization=62.5%
seq_length=33 -> block_table=[bs, 4*16] utilization=51.6%
seq_length=65 -> block_table=[bs, 8*16] utilization=50.8%

Now if instead of bucketing we allow the block_table to grow by one and we increase the block_size to 64:
seq_length=20 -> block_table=[bs, 1*32] utilization=62.5%
seq_length=33 -> block_table=[bs, 2*32] utilization=51.6%
seq_length=65 -> block_table=[bs, 3*32] utilization=67.7%

Reply:
> Ok, and why can't the block_table grow by one, exactly how it grows in non-graph runs? What's more, with bucketing it also grows. The only difference is that it grows by chunks > 1 .

In terms of current vLLM code - it cannot do that with CUDAGraphs/HPUGraphs, as block table will always be padded to the max. In terms of this PR - it can, there's nothing really preventing us from doing so, other than having huge amount of recorded graphs (with possibly large impact on host memory used).  If we want to pre-record all graphs with all block_table sizes (for each batch), we can do that with this bucketing solution. Here, you just need to adjust _BLOCK_COUNTS_TO_CAPTURE to be something like i for i in range(1, 4096)] and adjust `_get_graph_block_count` accordingly. You can overshoot with upper bound, as we'll skip capture of block counts extending beyond max_seq_len. 

> I agree that computing up to whole max_context_len makes no sense, and I don't see a reason why we should always pad to it.

Block table is an input tensor with dynamic shape, and its shape will impact how much you compute. We're tracing a CUDAGraph/HPUGraph with this particular tensor shape. I'm actually not sure what would happen if we passed a tensor with dynamic shape to recorded HPUGraph on a different shape. I assumed that we need to trace a graph with static tensors, and this is what we're doing as part of each bucket.

> Let's work with your example. Assuming block_size=16:
> 
> ```
> seq_length=20 -> block_table=[bs, 2*16] utilization=62.5%
> seq_length=33 -> block_table=[bs, 4*16] utilization=51.6%
> seq_length=65 -> block_table=[bs, 8*16] utilization=50.8%
> ```
> 
> Now if instead of bucketing we allow the block_table to grow by one and we increase the block_size to 32:
> 
> ```
> seq_length=20 -> block_table=[bs, 1*32] utilization=62.5%
> seq_length=33 -> block_table=[bs, 2*32] utilization=51.6%
> seq_length=65 -> block_table=[bs, 3*32] utilization=67.7%
> ```

That is all correct. The second one is better. I'll admit that 1, 2, 4, 8, 16, 24, 32, 40, ..., 256 buckets were pretty arbitrary and copy-pasted from batch bucketing, and we could definitely fine-tune it. On the other hand, you could make the same argument to batch bucketing, you'd have the same utilization rates. 

Without any context len bucketing, we have two options: either run it in lazy and adjust `max_num_blocks_per_seq` as you go, or run it with HPUGraphs and always overcompute for the highest possible value of `max_num_blocks_per_seq`. This patch gives us option to not do that.